cpp/src/gis/cuda/functor/geometry_output.impl.h

```diff
@@ -0,0 +1,100 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// NOTE:
+// this file is template of all functors with a Geometry output
+// such as ST_Point, ST_Envelope
+#pragma once
+#include "gis/cuda/common/gis_definitions.h"
+#include "gis/cuda/functor/geometry_output.h"
+
+namespace zilliz {
+namespace gis {
+namespace cuda {
+
+namespace internal {
+namespace geometry_output {
+using DataState = GeometryVector::DataState;
+
+template <typename Functor>
+__global__ void FillInfoKernel(Functor functor, GpuContext results) {
+  assert(results.data_state == DataState::FlatOffset_EmptyInfo);
+  auto index = threadIdx.x + blockIdx.x * blockDim.x;
+  if (index < results.size) {
+    OutputInfo out_info = functor(index, results, true);
+    results.tags[index] = out_info.tag;
+    results.meta_offsets[index] = out_info.meta_size;
+    results.value_offsets[index] = out_info.value_size;
+  }
+}
+
+DEVICE_RUNNABLE inline void AssertInfo(OutputInfo info, const GpuContext& ctx,
+                                       int index) {
+  assert(info.tag.data == ctx.get_tag(index).data);
```

Comment:
will these asserts lead to system crash?

Reply:
Yes but these asserts should be guaranteed as post-condition logical requirements. 

If triggered, they might not be capable of being handled by exceptions, since the GPU heap might have been corrupted.